crates/forge_domain/src/orch.rs

```diff
@@ -14,6 +16,34 @@ use crate::*;
 
 type ArcSender = Arc<tokio::sync::mpsc::Sender<anyhow::Result<AgentMessage<ChatResponse>>>>;
 
+// Maximum number of retry attempts for retryable operations
+const MAX_RETRY_ATTEMPTS: usize = 3;
+
+// Helper function that checks if an error is retriable and sends a retry event
+// if needed
+async fn check_and_log_retry_condition(
+    err: &anyhow::Error,
+    attempt: usize,
+    max_attempts: usize,
+    agent: Option<&Agent>,
+    sender: Option<&ArcSender>,
+) -> bool {
+    let is_retriable = RetryError::is_retriable(err) && attempt < max_attempts;
+    if let (true, Some(a), Some(s)) = (is_retriable, agent, sender) {
+        let _ = s
+            .send(Ok(AgentMessage {
+                agent: a.id.clone(),
+                message: ChatResponse::Retry {
+                    error: RetryError::from_error(err),
+                    attempt,
+                    max_attempts,
+                },
+            }))
+            .await;
+    }
+    is_retriable
+}
```

Comment:
Create a new struct that holds all the retry related information. Use simpler method calls so that easier work with.